File: test/features/billing/campaign_code_eligibility_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:wanote/features/billing/domain/campaign_code_models.dart';

CampaignCode _code({
  bool active = true,
  int maxRedemptions = 100,
  int redemptionCount = 0,
  String? referrerUid,
}) {
  return CampaignCode(
    code: 'SUMMER2026',
    active: active,
    maxRedemptions: maxRedemptions,
    redemptionCount: redemptionCount,
    referrerUid: referrerUid,
  );
}

void main() {
  group('CampaignCodeEligibility.evaluate', () {
    test('unknown code (no campaign_codes/{code} document): ineligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: null,
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.unknownCode,
      );
    });

    test('inactive code: ineligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(active: false),
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.inactive,
      );
    });

    test('active code, already redeemed by this user: ineligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(),
        alreadyRedeemedByUser: true,
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.alreadyRedeemedByUser,
      );
    });

    test('active code, under the redemption cap, not yet redeemed by this user: eligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(maxRedemptions: 100, redemptionCount: 50),
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionEligible>());
    });

    test('active code exactly at the redemption cap: ineligible (cap reached)', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(maxRedemptions: 100, redemptionCount: 100),
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.redemptionCapReached,
      );
    });

    test('active code over the redemption cap (redemptionCount > maxRedemptions): ineligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(maxRedemptions: 100, redemptionCount: 101),
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.redemptionCapReached,
      );
    });

    test('active code, one redemption remaining (under cap by exactly 1): eligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(maxRedemptions: 100, redemptionCount: 99),
        alreadyRedeemedByUser: false,
      );

      expect(result, isA<RedemptionEligible>());
    });

    test('referral code redeemed by its own referrer: ineligible (self-referral)', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(referrerUid: 'user-1'),
        alreadyRedeemedByUser: false,
        uid: 'user-1',
      );

      expect(result, isA<RedemptionIneligible>());
      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.selfReferral,
      );
    });

    test('referral code redeemed by someone other than its referrer: eligible', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(referrerUid: 'user-1'),
        alreadyRedeemedByUser: false,
        uid: 'user-2',
      );

      expect(result, isA<RedemptionEligible>());
    });

    test('inactive takes priority over redemption cap when both apply', () {
      final result = CampaignCodeEligibility.evaluate(
        campaignCode: _code(active: false, maxRedemptions: 1, redemptionCount: 1),
        alreadyRedeemedByUser: false,
      );

      expect(
        (result as RedemptionIneligible).reason,
        RedemptionIneligibleReason.inactive,
      );
    });
  });

  group('CampaignCode.hasRedemptionsRemaining', () {
    test('true when under the cap', () {
      expect(_code(maxRedemptions: 10, redemptionCount: 9).hasRedemptionsRemaining, isTrue);
    });

    test('false when exactly at the cap', () {
      expect(_code(maxRedemptions: 10, redemptionCount: 10).hasRedemptionsRemaining, isFalse);
    });

    test('false when over the cap', () {
      expect(_code(maxRedemptions: 10, redemptionCount: 11).hasRedemptionsRemaining, isFalse);
    });
  });

  group('ReferralCodeGenerator.deriveFrom', () {
    test('is deterministic for the same uid', () {
      final a = ReferralCodeGenerator.deriveFrom('abcdef1234567890');
      final b = ReferralCodeGenerator.deriveFrom('abcdef1234567890');
      expect(a, b);
    });

    test('differs for different uids', () {
      final a = ReferralCodeGenerator.deriveFrom('userAAAAAAAA');
      final b = ReferralCodeGenerator.deriveFrom('userBBBBBBBB');
      expect(a, isNot(b));
    });

    test('handles a uid shorter than the short-id length without throwing', () {
      expect(() => ReferralCodeGenerator.deriveFrom('abc'), returnsNormally);
    });
  });
}

File: functions/test/campaignCodeEligibility.test.ts
import { describe, expect, it } from 'vitest';

import { evaluateEligibility, type CampaignCode } from '../src/lib/campaignCodeEligibility';

/**
 * These rules decide whether a user gets a free month of premium, so they are
 * now enforced on the server -- the Dart copy in campaign_code_models.dart is
 * only there to give fast feedback. Before this moved, the endpoint granted
 * premium to anyone who asked, so these were effectively unenforced.
 *
 * The reason strings are part of the contract: the Flutter UI maps them onto
 * RedemptionIneligibleReason to pick a localized message, so renaming one
 * here silently degrades the app to a generic error.
 */

const code = (overrides: Partial<CampaignCode> = {}): CampaignCode => ({
  active: true,
  maxRedemptions: 10,
  redemptionCount: 0,
  referrerUid: null,
  ...overrides,
});

describe('evaluateEligibility', () => {
  it('allows a normal redemption', () => {
    expect(
      evaluateEligibility({ code: code(), alreadyRedeemedByUser: false, uid: 'u1' }),
    ).toEqual({ eligible: true });
  });

  it('rejects a code that does not exist', () => {
    expect(
      evaluateEligibility({ code: null, alreadyRedeemedByUser: false, uid: 'u1' }),
    ).toEqual({ eligible: false, reason: 'unknownCode' });
  });

  it('rejects a deactivated code', () => {
    expect(
      evaluateEligibility({
        code: code({ active: false }),
        alreadyRedeemedByUser: false,
        uid: 'u1',
      }),
    ).toEqual({ eligible: false, reason: 'inactive' });
  });

  it('rejects redeeming your own referral code', () => {
    expect(
      evaluateEligibility({
        code: code({ referrerUid: 'u1' }),
        alreadyRedeemedByUser: false,
        uid: 'u1',
      }),
    ).toEqual({ eligible: false, reason: 'selfReferral' });
  });

  it("does not mistake someone else's referral code for self-referral", () => {
    expect(
      evaluateEligibility({
        code: code({ referrerUid: 'u2' }),
        alreadyRedeemedByUser: false,
        uid: 'u1',
      }),
    ).toEqual({ eligible: true });
  });

  it('rejects a second redemption by the same user', () => {
    expect(
      evaluateEligibility({ code: code(), alreadyRedeemedByUser: true, uid: 'u1' }),
    ).toEqual({ eligible: false, reason: 'alreadyRedeemedByUser' });
  });

  it('rejects once the cap is reached', () => {
    expect(
      evaluateEligibility({
        code: code({ maxRedemptions: 3, redemptionCount: 3 }),
        alreadyRedeemedByUser: false,
        uid: 'u1',
      }),
    ).toEqual({ eligible: false, reason: 'redemptionCapReached' });
  });

  it('rejects a count that somehow exceeded the cap', () => {
    // Defensive: a stored count above the cap must still block, not wrap
    // around into "capacity remaining".
    expect(
      evaluateEligibility({
        code: code({ maxRedemptions: 3, redemptionCount: 9 }),
        alreadyRedeemedByUser: false,
        uid: 'u1',
      }),
    ).toEqual({ eligible: false, reason: 'redemptionCapReached' });
  });

  it('reports self-referral ahead of an already-redeemed marker', () => {
    // Both apply; the more specific reason is the more useful message.
    expect(
      evaluateEligibility({
        code: code({ referrerUid: 'u1' }),
        alreadyRedeemedByUser: true,
        uid: 'u1',
      }),
    ).toEqual({ eligible: false, reason: 'selfReferral' });
  });
});
